### PR TITLE
Convert hierarchy to Fragment

### LINKED ISSUE #6303: Update summary/jump icon
Blocked by #5420

We've heard feedback that the [icon isn't intuitive](https://www.notion.so/getodk/Jump-menu-icon-is-confusing-efecfe9a11f84f469088c59531029696?pvs=4). 

We reviewed [different options](https://www.figma.com/design/KFi9hIQdrRgo0rLqIC7Sg8/ODK-user-flows?node-id=6228-89614&t=CbI3Km7LacH9cKvO-1) as a team, but landed on the [classic list icon](https://www.figma.com/design/KFi9hIQdrRgo0rLqIC7Sg8/ODK-user-flows?node-id=6228-20698&t=CbI3Km7LacH9cKvO-1) because it's the easiest to recognize, and hopefully won't need explanation. 

- Update east arrow icon with [list icon ](https://fonts.google.com/icons?selected=Material+Symbols+Outlined:list:FILL@0;wght@400;GRAD@0;opsz@24&icon.query=list&icon.size=24&icon.color=%235f6368). 
- We are still keeping the up icon when the user is drilling down into groups or repeats. 
- We will need to talk about this change in advance (maybe Insiders and a forum post).

### PR DESCRIPTION
**Pull request by seadowg**
Work towards #5420
Closes #6303

#### Why is this the best possible solution? Were any other approaches considered?

I've tried to keep changes to a minimal here with the main goal just being to get the majority of the hierarchy code into a `Fragment`. There's only two structural changes to highlight really:

1. I ended up rewriting the hierarchy's options menu so that it uses the new `MenuProvider` API. I felt like this made more sense than rewriting it to use the deprecated `Fragment` methods.
2. The new Fragment uses a new `FormHierarchyViewModel` to manage state. This will most likely get merged into `FormEntryViewModel`, but I think that'll make more sense once form entry has also been moved to a Fragment.

I also added some new tests for code that I found had none.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

This shouldn't change any behaviour and the big risk here is that it either does or introduces bugs to flows using the form hierarchy. A general test of using the hierarchy (during form entry and viewing) would be good here.

This also changes the jump/summary icon to the new one!

#### Before submitting this PR, please make sure you have:
- [x] added or modified tests for any new or changed behavior
- [x] run `./gradlew connectedAndroidTest` (or `./gradlew testLab`) and confirmed all checks still pass
- [x] added a comment above any new strings describing it for translators
- [x] added any new strings with date formatting to `DateFormatsTest`
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by seadowg** [comment]
@getodk/testers marking this ready for testing before review as it'd be good to get a first pass before I clean things up in case there are some fundamental problems here I've missed.

**Comment by dbemke** [comment]
There used to be 2 icons: to go to the hierarchy view and to go "up" in the hierarchy (the second screenshot). Now there's only one icon and it feels a bit confusing in groups and repeats cause the new icon doesn't suggest where the user is moved after tapping it. Is it expected?
newHierarchy
hierarchyIcon

**Comment by seadowg** [comment]
> There used to be 2 icons: to go to the hierarchy view and to go "up" in the hierarchy (the second screenshot). Now there's only one icon and it feels a bit confusing in groups and repeats cause the new icon doesn't suggest where the user is moved after tapping it. Is it expected?

I think this was the intention of https://github.com/getodk/collect/issues/6303, but please correct me if I'm wrong @alyblenkin.

**Comment by alyblenkin** [comment]
> I think this was the intention of https://github.com/getodk/collect/issues/6303, but please correct me if I'm wrong @alyblenkin.

This is a good catch! The goal was just to update the east arrow icon, not the up icon as well. I think the up arrow is still a helpful visual indicator that you're going back up to the table of contents.

**Comment by seadowg** [comment]
> This is a good catch! The goal was just to update the east arrow icon, not the up icon as well. I think the up arrow is still a helpful visual indicator that you're going back up to the table of contents.

That makes sense. Just to be sure though before I unwind anything, [the Figma link](https://www.figma.com/design/KFi9hIQdrRgo0rLqIC7Sg8/ODK-user-flows?node-id=6228-20698&t=CbI3Km7LacH9cKvO-1) shows the hierarchy with the list icon in the top right:

Screenshot 2024-12-16 at 16 47 05

I'm guessing that's a mistake?

**Comment by alyblenkin** [comment]
Sorry I think the group folder icon in the first mockup is a mistake. However, now I'm second guessing myself because we discussed this a while back. @lognaturel correct me if I'm wrong here, but our goal was just update the east arrow, not the up arrow icon as well.

**Comment by alyblenkin** [comment]
To clarify, I think the list icon would work on it's own if we didn't already have the up arrow, but I think it's a pretty ingrained pattern now. I think the list icon would suggest you go back to the top level given the current pattern where there is more of this concept of moving up or down, but you could also imagine tapping to keep going to the main table of contents. I don't think this is a stronger design choice, rather it could be another option. Not worth considering given we haven't had folks bring up any issues.

**Comment by lognaturel** [comment]
Feels right to me to leave the up arrow as it is. I think navigating up a level feels relatively clear. I don't think the table of contents icon works well there because it would suggest you go all the way back to the top level.

**Comment by alyblenkin** [comment]
@seadowg - sorry that you have to unwind things there! I've update the design.

**Comment by seadowg** [comment]
> sorry that you have to unwind things there! I've update the design.

No worries! I just wanted to double check.

**Comment by seadowg** [comment]
@getodk/testers that should be good to go again!

**Comment by dbemke** [comment]
**Tested with success!**

Verified on devices with Android 10, 14

**Verified cases:**
- the new icon in the hierarchy
- groups and repeats
- the hierarchy view in new forms, drafts, Ready to send, Sent 
- the hierarchy view after recovering a savepoint
- moving backwards enabled and disabled
- RTL
- light and dark  mode
- don't keep activities enabled and disabled

File: collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormHierarchyTest.java [hunk 1]
```diff
@@ -238,4 +240,33 @@ public void theListOfQuestionsShouldBeScrolledToTheLastDisplayedQuestionAfterOpe
                 .assertTextDoesNotExist("t1")
                 .assertTextDoesNotExist("t2");
     }
+
+    @Test
+    public void whenViewFormInHierarchyForRepeat_noDeleteButtonAppears() {
```
Comment (grzesiek2010) [inline collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormHierarchyTest.java]: Above you renamed `whenInHierarchyForRepeat_clickingPlus_addsRepeatAtEndOfSeries` adding `Group` so for consistency here the test also should be named `whenViewFormInHierarchyForRepeatGroup_noDeleteButtonAppears`.

File: collect_app/src/main/java/org/odk/collect/android/formhierarchy/FormHierarchyFragment.java [hunk 2]
```diff
@@ -42,261 +35,127 @@
...
-
-    /**
-     * The toolbar menu.
-     */
-    private Menu optionsMenu;
-
-    protected Button jumpBeginningButton;
-    protected Button jumpEndButton;
-    protected RecyclerView recyclerView;
+public class FormHierarchyFragment extends Fragment {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/formhierarchy/FormHierarchyFragment.java:57]: Are we planning to convert this to Kotlin in a follow-up pull request?

File: collect_app/src/main/res/layout/form_hierarchy_layout.xml [hunk 3]
```diff
@@ -0,0 +1,100 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (grzesiek2010) [inline collect_app/src/main/res/layout/form_hierarchy_layout.xml:2]: Now `hierarchy_layout` is redundant and you can remove it.

File: collect_app/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -328,6 +326,8 @@ the specific language governing permissions and limitations under the License.
                 <data android:mimeType="vnd.android.cursor.item/vnd.odk.instance" />
             </intent-filter>
         </activity>
+        <activity android:name="org.odk.collect.android.formhierarchy.FormHierarchyFragmentHostActivity" />
+        <activity android:name="org.odk.collect.android.formhierarchy.FormHierarchyActivity" />
```
Comment (grzesiek2010) [inline collect_app/src/main/AndroidManifest.xml]: This should be removed.

File: collect_app/src/main/java/org/odk/collect/android/formhierarchy/FormHierarchyFragment.java [hunk 5]
```diff
@@ -42,261 +35,127 @@
...
-import org.odk.collect.strings.localization.LocalizedActivity;
 
 import java.util.ArrayList;
 import java.util.List;
 
-import javax.inject.Inject;
-
 import timber.log.Timber;
 
-public class FormHierarchyActivity extends LocalizedActivity implements DeleteRepeatDialogFragment.DeleteRepeatDialogCallback {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/formhierarchy/FormHierarchyFragment.java]: There are still a couple of comments where `FormHierarchyActivity` is used instead of `FormHierarchyFragmentHostActivity` plus in the `STATE.md` file.

File: icons/src/main/res/drawable/ic_baseline_list_24.xml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
+    android:height="24dp"
+    android:autoMirrored="true"
+    android:tint="?colorOnSurface"
+    android:viewportWidth="24"
+    android:viewportHeight="24">
+    <path
+        android:fillColor="?colorSurface"
```
Comment (grzesiek2010) [inline icons/src/main/res/drawable/ic_baseline_list_24.xml]: Why not `colorOnSurface`?

File: collect_app/src/main/java/org/odk/collect/android/formentry/repeats/DeleteRepeatDialogFragment.java [hunk 7]
```diff
@@ -64,7 +64,12 @@ public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
             if (i == BUTTON_POSITIVE) { // yes
                 formController.getAuditEventLogger().logEvent(AuditEvent.AuditEventType.DELETE_REPEAT, true, System.currentTimeMillis());
                 formController.deleteRepeat();
-                callback.deleteGroup();
+
+                if (callback != null) {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/formentry/repeats/DeleteRepeatDialogFragment.java]: Can't we have one way of communication?

File: collect_app/src/main/java/org/odk/collect/android/formhierarchy/FormHierarchyFragment.java [hunk 8]
```diff
@@ -42,261 +35,127 @@
...
+    public void onAttach(@NonNull Context context) {
+        super.onAttach(context);
 
-    @Inject
-    public AutoSendSettingsProvider autoSendSettingsProvider;
+        formEntryViewModel = new ViewModelProvider(requireActivity(), viewModelFactory).get(FormEntryViewModel.class);
+        formHierarchyViewModel = new ViewModelProvider(this, new ViewModelProvider.Factory() {
+            @NonNull
+            @Override
+            public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/formhierarchy/FormHierarchyFragment.java:82]: We could move this to `FormEntryViewModelFactory` and just use `viewModelFactory` as above.

File: collect_app/src/main/java/org/odk/collect/android/formhierarchy/FormHierarchyFragment.java [hunk 9]
```diff
@@ -603,21 +175,25 @@ public void refreshView() {
...
         try {
             FormController formController = formEntryViewModel.getFormController();
 
             // Save the current index so we can return to the problematic question
             // in the event of an error.
-            currentIndex = formController.getFormIndex();
+            formHierarchyViewModel.setCurrentIndex(formController.getFormIndex());
 
-            elementsToDisplay = new ArrayList<>();
+            ArrayList<HierarchyItem> elementsToDisplay = new ArrayList<>();
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/formhierarchy/FormHierarchyFragment.java]: You should use the `List` interface here:
`List<HierarchyItem> elementsToDisplay = new ArrayList<>();
`

File: collect_app/src/main/java/org/odk/collect/android/formhierarchy/FormHierarchyFragment.java [hunk 10]
```diff
@@ -792,13 +370,14 @@ private void refreshView(boolean isGoingUp) {
                 event = formController.stepToNextEvent(JavaRosaFormController.STEP_INTO_GROUP);
             }
 
-            recyclerView.setAdapter(new HierarchyListAdapter(elementsToDisplay, this::onElementClick));
+            formHierarchyViewModel.setElementsToDisplay(elementsToDisplay);
+            recyclerView.setAdapter(new HierarchyListAdapter(formHierarchyViewModel.getElementsToDisplay(), this::onElementClick));
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/formhierarchy/FormHierarchyFragment.java]: I think here you can directly use `elementsToDisplay`. However, if you always access the list via the ViewModel please update it below in line `formController.jumpToIndex(elementsToDisplay.get(0).getFormIndex());` as well.

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]

File: collect_app/src/main/java/org/odk/collect/android/formhierarchy/FormHierarchyFragment.java (hunk 2 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/formhierarchy/FormHierarchyFragment.java:57]: Not in the short term. I wanted to stay away from bulk Kotlin converts here to keep the diffs tight and limit any extra work. My strategy was to just keep the Java contained in `FormHierarchyFragment` and try and make as little changes there as possible.

File: icons/src/main/res/drawable/ic_baseline_list_24.xml (hunk 6 above)
Comment (seadowg) [reply icons/src/main/res/drawable/ic_baseline_list_24.xml]: I'm pretty sure they're flipped from what you'd think here: the tint is `colorOnSurface` and the path is `colorSurface`. Everything looks correct right?

File: collect_app/src/main/java/org/odk/collect/android/formhierarchy/FormHierarchyFragment.java (hunk 8 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/formhierarchy/FormHierarchyFragment.java:82]: `FormHierarchyViewModel` is currently private to `FormHierarchyFragment` and I reckon as we continue we'll probably remove it and move its functionality into `FormEntryViewModel`.

File: icons/src/main/res/drawable/ic_baseline_list_24.xml (hunk 6 above)
Comment (grzesiek2010) [reply icons/src/main/res/drawable/ic_baseline_list_24.xml]: But why not to use `colorOnSurface` in both as we do in every other icon? What's the point of using `colorSurface` here?

Comment (seadowg) [reply icons/src/main/res/drawable/ic_baseline_list_24.xml]: Good question! It looks like we just don't need `tint`. I'd always get confused by this, but the tint color [is "blended" in with the others](https://developer.android.com/reference/android/R.attr#tint) so here the `path` was being rendered with `colorSurface` but then it would just be "blended" (just overridden in this case) with `colorOnSurface`. If I remove `tint` and set the path to `colorOnSurface` it seems to work as expected.

Comment (grzesiek2010) [reply icons/src/main/res/drawable/ic_baseline_list_24.xml]: > If I remove tint and set the path to colorOnSurface it seems to work as expected.

Yes, but when you change themes it might not work as expected see this pr https://github.com/getodk/collect/pull/5537
Maybe now as removed changing the theme in the app it is no longer needed but I'm not sure. If you want to remove it please make sure it is not needed or maybe a better idea is to keep it and file an issue to investigate if we still need it.

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]
https://github.com/getodk/collect/pull/6516#discussion_r1907155873

**Review by grzesiek2010 — APPROVED** [review]
